File: src/utils/subgraph_utils.cpp
#include "utils/subgraph_utils.h"
#include "utils/time_utils.h"
#include <curl/curl.h>
#include <nlohmann/json.hpp>
#include <string>
#include <vector>
#include <optional>
#include <iostream>
#include <fstream>
#include <algorithm>

using json = nlohmann::json;

size_t WriteCallback_subgraph_utils(void* contents, size_t size, size_t nmemb, std::string* buffer) {
    size_t totalSize = size * nmemb;
    buffer->append((char*)contents, totalSize);
    return totalSize;
}

std::optional<json> graphql_post(
    const std::string& apiSubgraphs,
    const std::string& idSubgraphs,
    const std::string& graphqlQuery
) {
    CURL* curl;
    CURLcode res;
    std::string readBuffer;

    json payload = {
        {"query", graphqlQuery},
        {"operationName", "Subgraphs"},
        {"variables", {}}
    };
    std::string postData = payload.dump();

    curl = curl_easy_init();
    if (curl) {
        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        std::string authHeader = "Authorization: Bearer " + apiSubgraphs;
        headers = curl_slist_append(headers, authHeader.c_str());

        std::string url = "https://gateway.thegraph.com/api/subgraphs/id/" + idSubgraphs;

        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postData.c_str());
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback_subgraph_utils);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);

        res = curl_easy_perform(curl);

        long http_code = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);

        curl_easy_cleanup(curl);
        curl_slist_free_all(headers);

        if (res != CURLE_OK || http_code != 200) {
            std::cerr << "cURL error: " << curl_easy_strerror(res) << std::endl;
            return std::nullopt;
        }

        try {
            return json::parse(readBuffer);
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return std::nullopt;
}

std::optional<PriceDatetimeSeries> parse_price_datetime_series(const json& json_data) {
    try {
        PriceDatetimeSeries result;
        std::string token0_symbol, token1_symbol;
        if (json_data.contains("data") && json_data["data"].contains("pool")) {
            const auto& pool = json_data["data"]["pool"];
            token0_symbol = pool["token0"]["symbol"].get<std::string>();
            token1_symbol = pool["token1"]["symbol"].get<std::string>();
        }
        if (json_data.contains("data") && json_data["data"].contains("poolHourDatas")) {
            for (const auto& hour_data : json_data["data"]["poolHourDatas"]) {
                time_t timestamp = hour_data["periodStartUnix"].get<time_t>();
                result.datetime.push_back(unix_to_datetime(timestamp));
                result.price.push_back(std::stod(hour_data["token0Price"].get<std::string>()));
            }
        }
        // Optionally, you could store token0_symbol/token1_symbol in result if needed
        return result;
    } catch (const std::exception& e) {
        std::cerr << "Error parsing price/datetime series: " << e.what() << std::endl;
        return std::nullopt;
    }
}

std::optional<PriceDatetimeSeries> fetch_price_datetime_series(
    const std::string& apiSubgraphs,
    const std::string& idSubgraphs,
    const std::string& poolAddress,
    const std::string& startDate,
    const std::string& endDate
) {
    time_t start_ts = date_to_unix(startDate);
    time_t end_ts = date_to_unix(endDate) + 86399;

    // Query both pool info (for token0/token1 symbols) and poolHourDatas
    std::string graphqlQuery = R"({
    pool(id: ")" + poolAddress + R"(") {
        token0 { symbol }
        token1 { symbol }
    }
    poolHourDatas(
        orderBy: periodStartUnix,
        orderDirection: asc,
        where: { pool: ")" + poolAddress + R"(", periodStartUnix_gte: )" + std::to_string(start_ts) + R"(, periodStartUnix_lte: )" + std::to_string(end_ts) + R"( }
    ) {
        periodStartUnix
        token0Price
    }
    })";

    auto json_data_opt = graphql_post(apiSubgraphs, idSubgraphs, graphqlQuery);
    if (!json_data_opt.has_value()) return std::nullopt;
    return parse_price_datetime_series(json_data_opt.value());
}

// Helper to load pool addresses from JSON file
std::unordered_map<std::string, std::unordered_map<std::string, std::string>> load_pool_addresses(const std::string& json_path) {
    std::unordered_map<std::string, std::unordered_map<std::string, std::string>> pool_map;
    std::ifstream file(json_path);
    if (!file.is_open()) {
        std::cerr << "[DEBUG] Could not open pool addresses JSON file: " << json_path << std::endl;
        std::string abs_path = "../" + json_path;
        file.open(abs_path);
        if (!file.is_open()) {
            std::cerr << "[DEBUG] Could not open fallback pool addresses JSON file: " << abs_path << std::endl;
            std::string just_filename = "v3_pool_addresses.json";
            file.open(just_filename);
            if (!file.is_open()) {
                std::cerr << "[DEBUG] Could not open fallback pool addresses JSON file: " << just_filename << std::endl;
                return pool_map;
            }
        }
    }
    nlohmann::json j;
    try {
        file >> j;
    } catch (const std::exception& e) {
        std::cerr << "[DEBUG] Failed to parse JSON: " << e.what() << std::endl;
        return pool_map;
    }
    for (auto it = j.begin(); it != j.end(); ++it) {
        std::string subgraph_id = it.key();
        std::transform(subgraph_id.begin(), subgraph_id.end(), subgraph_id.begin(), ::tolower);
        if (it->is_object()) {
            for (auto pit = it->begin(); pit != it->end(); ++pit) {
                std::string pool_key = pit.key();
                std::transform(pool_key.begin(), pool_key.end(), pool_key.begin(), ::toupper);
                std::string address = pit.value();
                std::transform(address.begin(), address.end(), address.begin(), ::tolower);
                pool_map[subgraph_id][pool_key] = address;
            }
        }
    }
    return pool_map;
}

// Helper to get pool address for a currency symbol pair (e.g., "ETH_USDC" or "USDC_ETH") and subgraph id
std::string get_pool_address(const std::string& symbol_pair, const std::string& subgraph_id) {
    static auto pool_map = load_pool_addresses("utils/v3_pool_addresses.json");
    // Normalize to uppercase for pool key and lowercase for subgraph id
    std::string key = symbol_pair;
    std::transform(key.begin(), key.end(), key.begin(), ::toupper);
    std::string sub_id = subgraph_id;
    std::transform(sub_id.begin(), sub_id.end(), sub_id.begin(), ::tolower);
    auto sub_it = pool_map.find(sub_id);
    if (sub_it == pool_map.end()) {
        return "";
    }
    auto& pools = sub_it->second;
    auto it = pools.find(key);
    if (it != pools.end()) {
        return it->second;
    }
    // Try reversed pair in uppercase
    size_t sep = key.find('_');
    if (sep != std::string::npos) {
        std::string reversed = key.substr(sep + 1) + "_" + key.substr(0, sep);
        it = pools.find(reversed);
        if (it != pools.end()) {
            return it->second;
        }
    }
    return "";
}

File: python/utils/subgraph_utils.py
import requests

import pandas as pd

from datetime import datetime

from concurrent.futures import ThreadPoolExecutor, as_completed

def date_to_unix(date_str):
    """
    Convert a date string in 'YYYY-MM-DD' format to a Unix timestamp (seconds since epoch).
    Args:
        date_str (str): Date string in 'YYYY-MM-DD' format.
    Returns:
        int: Unix timestamp.
    """
    dt = datetime.strptime(date_str, '%Y-%m-%d')
    return int(dt.timestamp())

def make_price_query(pool_address, start_ts, end_ts):
    """
    Create a GraphQL query string for fetching hourly price, volumeUSD, and liquidity for a single Uniswap V3 pool.
    Args:
        pool_address (str): Pool address.
        start_ts (int): Start timestamp (Unix).
        end_ts (int): End timestamp (Unix).
    Returns:
        str: GraphQL query string.
    """
    return f"""
    {{
      poolHourDatas(first: 1000, where: {{pool: \"{pool_address}\", periodStartUnix_gte: {start_ts}, periodStartUnix_lte: {end_ts}}}, orderBy: periodStartUnix, orderDirection: asc) {{
        periodStartUnix
        token0Price
        volumeUSD
        liquidity
      }}
    }}
    """

def make_price_query_multi(pool_addresses, start_ts, end_ts):
    """
    Create a GraphQL query string for fetching hourly price, volumeUSD, and liquidity for multiple Uniswap V3 pools.
    Args:
        pool_addresses (list): List of pool addresses.
        start_ts (int): Start timestamp (Unix).
        end_ts (int): End timestamp (Unix).
    Returns:
        str: GraphQL query string.
    """
    pool_list = ', '.join(f'"{addr}"' for addr in pool_addresses)
    return f"""
{{
  poolHourDatas(
    first: 1000,
    where: {{
      pool_in: [ {pool_list} ]
      periodStartUnix_gte: {start_ts}
      periodStartUnix_lte: {end_ts}
    }},
    orderBy: periodStartUnix,
    orderDirection: asc
  ) {{
    pool {{
      id
    }}
    periodStartUnix
    token0Price
    volumeUSD
    liquidity
  }}
}}
"""

def fetch_pool_hourly_data(api_key, subgraph_id, pool_address, start_date, end_date):
    """
    Fetch hourly price (token0Price as 'price'), volumeUSD, liquidity, and datetimes for a Uniswap V3 pool from the subgraph.
    Args:
        api_key (str): The Graph API key.
        subgraph_id (str): Subgraph ID.
        pool_address (str): Pool address.
        start_date (str): Start date in 'YYYY-MM-DD' format.
        end_date (str): End date in 'YYYY-MM-DD' format.
    Returns:
        pd.DataFrame: DataFrame with columns periodStartUnix, price, volumeUSD, liquidity.
    """
    start_ts = date_to_unix(start_date)
    end_ts = date_to_unix(end_date)
    all_data = []
    last_ts = start_ts
    while True:
        graphql_query = make_price_query(pool_address, last_ts, end_ts)
        payload = {
            "query": graphql_query,
            "operationName": "Subgraphs",
            "variables": {}
        }
        url = f"https://gateway.thegraph.com/api/subgraphs/id/{subgraph_id}"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}"
        }
        response = requests.post(url, json=payload, headers=headers)
        if response.status_code != 200:
            raise Exception(f"Query failed with status code {response.status_code}: {response.text}")
        result = response.json()
        hour_data = result['data']['poolHourDatas']
        if not hour_data:
            break
        all_data.extend(hour_data)
        if len(hour_data) < 1000:
            break
        # Update last_ts to one after the last returned timestamp to avoid overlap
        last_ts = int(hour_data[-1]['periodStartUnix']) + 1

    df = pd.DataFrame(all_data)
    if not df.empty:
        df['periodStartUnix'] = pd.to_datetime(df['periodStartUnix'], unit='s')
        df = df.rename(columns={"token0Price": "price"})
        df['price'] = pd.to_numeric(df['price'], errors='coerce').astype(float)
        df['volumeUSD'] = pd.to_numeric(df['volumeUSD'], errors='coerce').astype(float)
        df['liquidity'] = pd.to_numeric(df['liquidity'], errors='coerce').astype(float)
    return df

def fetch_pools_hourly_data_batched(api_key, subgraph_id, pool_addresses, start_date, end_date):
    """
    Fetch hourly price, volumeUSD, and liquidity for multiple pools in sequential batches.

    Args:
        api_key (str): The Graph API key.
        subgraph_id (str): Subgraph ID.
        pool_addresses (list): List of pool addresses (lowercase strings).
        start_date (str): Start date in 'YYYY-MM-DD' format.
        end_date (str): End date in 'YYYY-MM-DD' format.

    Returns:
        dict: Mapping pool_address -> DataFrame of hourly data.
    """
    start_ts = date_to_unix(start_date)
    end_ts = date_to_unix(end_date)
    batch_size = 10  # Number of pools per batch
    pool_dict = {addr: [] for addr in pool_addresses}

    for batch_start in range(0, len(pool_addresses), batch_size):
        batch = pool_addresses[batch_start:batch_start + batch_size]
        all_data = []
        last_ts = start_ts

        while True:
            graphql_query = make_price_query_multi(batch, last_ts, end_ts)
            payload = {
                "query": graphql_query
            }
            url = f"https://gateway.thegraph.com/api/subgraphs/id/{subgraph_id}"
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}"
            }

            response = requests.post(url, json=payload, headers=headers)
            if response.status_code != 200:
                print(f"[ERROR] Query failed for batch {batch} with status code {response.status_code}: {response.text}")
                break

            result = response.json()
            hour_data = result.get("data", {}).get("poolHourDatas", [])
            if not hour_data:
                break

            all_data.extend(hour_data)
            if len(hour_data) < 1000:
                break
            last_ts = int(hour_data[-1]['periodStartUnix']) + 1

        # Group results by pool address
        for row in all_data:
            pool_obj = row.get('pool', {})
            addr = pool_obj.get('id')
            if not addr:
                # print(f"[WARNING] Missing pool ID in row: {row}")
                continue
            if addr not in pool_dict:
                # print(f"[WARNING] Pool address {addr} not in pool_dict (skipping)")
                continue
            pool_dict[addr].append(row)

        print(f"[fetch_pools_hourly_data_batched] Fetched {len(all_data)} rows for batch {batch_start // batch_size + 1}/{(len(pool_addresses) + batch_size - 1) // batch_size}")

    # Convert to DataFrames
    for addr, rows in pool_dict.items():
        if not isinstance(rows, list) or not rows:
            df = pd.DataFrame()
        else:
            df = pd.DataFrame(rows)
            if not df.empty:
                df['periodStartUnix'] = pd.to_datetime(df['periodStartUnix'], unit='s')
                df = df.rename(columns={"token0Price": "price"})
                df['price'] = pd.to_numeric(df['price'], errors='coerce').astype(float)
                df['volumeUSD'] = pd.to_numeric(df['volumeUSD'], errors='coerce').astype(float)
                df['liquidity'] = pd.to_numeric(df['liquidity'], errors='coerce').astype(float)
        pool_dict[addr] = df

    return pool_dict

def fetch_pools_hourly_data_batched_parallel(api_key, subgraph_id, pool_addresses, start_date, end_date, max_workers=8):
    """
    Fetch hourly price, volumeUSD, and liquidity for multiple pools using asynchronous batch requests.

    Args:
        api_key (str): The Graph API key.
        subgraph_id (str): Subgraph ID.
        pool_addresses (list): List of pool addresses (lowercase strings).
        start_date (str): Start date in 'YYYY-MM-DD' format.
        end_date (str): End date in 'YYYY-MM-DD' format.
        max_workers (int): Maximum number of threads for async requests.

    Returns:
        dict: Mapping pool_address -> DataFrame of hourly data.
    """
    start_ts = date_to_unix(start_date)
    end_ts = date_to_unix(end_date)
    batch_size = 10
    pool_dict = {addr: [] for addr in pool_addresses}

    def fetch_batch(batch, batch_idx):
        batch_data = []
        last_ts = start_ts
        while True:
            graphql_query = make_price_query_multi(batch, last_ts, end_ts)
            payload = {
                "query": graphql_query
            }
            url = f"https://gateway.thegraph.com/api/subgraphs/id/{subgraph_id}"
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}"
            }
            response = requests.post(url, json=payload, headers=headers)
            if response.status_code != 200:
                print(f"[ERROR] Batch {batch} failed: {response.status_code} {response.text}")
                return []
            result = response.json()
            hour_data = result.get("data", {}).get("poolHourDatas", [])
            if not hour_data:
                break
            batch_data.extend(hour_data)
            if len(hour_data) < 1000:
                break
            last_ts = int(hour_data[-1]["periodStartUnix"]) + 1
        print(f"[fetch_pools_hourly_data_multi] Fetched {len(batch_data)} rows for batch {batch_idx+1}/{(len(pool_addresses) + batch_size - 1) // batch_size}")
        return batch_data

    futures = []
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        for batch_start in range(0, len(pool_addresses), batch_size):
            batch = pool_addresses[batch_start:batch_start + batch_size]
            batch_idx = batch_start // batch_size
            futures.append(executor.submit(fetch_batch, batch, batch_idx))
        for fut in as_completed(futures):
            batch_result = fut.result()
            # print(f"[fetch_pools_hourly_data_multi] Received batch result with {len(batch_result)} rows")
            for row in batch_result:
                pool_obj = row.get("pool", {})
                addr = pool_obj.get("id")
                if not addr:
                    # print(f"[fetch_pools_hourly_data_multi] WARNING: Missing pool ID in row: {row}")
                    continue
                if addr not in pool_dict:
                    # print(f"[fetch_pools_hourly_data_multi] WARNING: Pool address {addr} not in pool_dict")
                    continue
                pool_dict[addr].append(row)

    # Convert to DataFrames
    # print("[fetch_pools_hourly_data_multi] Converting results to DataFrames...")
    for addr, rows in pool_dict.items():
        df = pd.DataFrame(rows)
        # print(f"[fetch_pools_hourly_data_multi] Pool {addr} has {len(df)} rows (empty={df.empty})")
        if not df.empty:
            df['periodStartUnix'] = pd.to_datetime(df['periodStartUnix'], unit='s')
            df = df.rename(columns={"token0Price": "price"})
            df['price'] = pd.to_numeric(df['price'], errors='coerce')
            df['volumeUSD'] = pd.to_numeric(df['volumeUSD'], errors='coerce')
            df['liquidity'] = pd.to_numeric(df['liquidity'], errors='coerce')
        pool_dict[addr] = df

    return pool_dict


def fetch_all_pool_addresses(api_key, subgraph_id, pool_tier=None, first=1000):
    """
    Fetch all pool addresses from a Uniswap subgraph, with optional filtering by pool tier.
    Args:
        api_key (str): The Graph API key.
        subgraph_id (str): Subgraph ID.
        pool_tier (str, optional): Pool tier ('LOW', 'MEDIUM', 'HIGH') or None for no filter.
        first (int): Number of pools to fetch per request.
    Returns:
        list: List of pool addresses (str).
    """
    pool_addresses = []
    skip = 0
    while True:
        filters = []
        if pool_tier:
            filters.append(f'tier: "{pool_tier.upper()}"')
        where_clause = (', '.join(filters))
        where_str = f'where: {{{where_clause}}}' if where_clause else ''
        query = f"""
        {{
          pools(first: {first}, skip: {skip} {',' if where_str else ''}{where_str}) {{
            id
          }}
        }}
        """
        payload = {
            "query": query,
            "operationName": "Subgraphs",
            "variables": {}
        }
        url = f"https://gateway.thegraph.com/api/subgraphs/id/{subgraph_id}"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}"
        }
        response = requests.post(url, json=payload, headers=headers)
        if response.status_code != 200:
            raise Exception(f"Query failed with status code {response.status_code}: {response.text}")
        result = response.json()
        if 'data' not in result or 'pools' not in result['data']:
            raise Exception(f"Unexpected response format: {result}")
        pools = result['data']['pools']
        if not pools:
            break
        pool_addresses.extend([p['id'] for p in pools])
        if len(pools) < first:
            break
        skip += first
    return pool_addresses